From 34d1e922f1a4369cbcad3d9a910ea1a980cd5503 Mon Sep 17 00:00:00 2001
From: Holmes Conan <holmesconan@gmail.com>
Date: Mon, 25 Oct 2021 09:16:36 +0000
Subject: [PATCH] fix: android cross compiling standard libraries

---
 setup.py | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/setup.py b/setup.py
index 94ee770851..b252535e46 100644
--- a/setup.py
+++ b/setup.py
@@ -689,7 +689,7 @@ class PyBuildExt(build_ext):
                     for line in fp.readlines():
                         if line.startswith("gcc version"):
                             is_gcc = True
-                        elif line.startswith("clang version"):
+                        elif "clang version" in line:
                             is_clang = True
                         elif line.startswith("#include <...>"):
                             in_incdirs = True
@@ -707,6 +707,18 @@ class PyBuildExt(build_ext):
         finally:
             os.unlink(tmpfile)
 
+        for opt in os.environ.get('LDSHARED', '').split():
+            if opt.startswith('-L'):
+                lib_dir = opt[2:]
+                add_dir_to_list(self.compiler.library_dirs, lib_dir)
+
+                inc_dir = os.path.join(lib_dir, '..', 'include')
+                if os.path.exists(inc_dir):
+                    add_dir_to_list(self.compiler.include_dirs, os.path.realpath(inc_dir))
+                inc_dir = os.path.join(lib_dir, '..', '..', 'include')
+                if os.path.exists(inc_dir):
+                    add_dir_to_list(self.compiler.include_dirs, os.path.realpath(inc_dir))
+
     def add_ldflags_cppflags(self):
         # Add paths specified in the environment variables LDFLAGS and
         # CPPFLAGS for header and library files.
-- 
2.25.1

